pennylane/transforms/sign_expand.py

```diff
@@ -0,0 +1,327 @@
+        elif gate == "Y":
+            ops.append(qml.RX(-np.pi / 2, wires=[wire]))
+    return ops
+
+
+def MultiCRZ(theta, wires, control):
+    """
+    Implements a controlled decomposition of qml.MultiRZ
+
+    Args:
+        theta (float): rotation angle :math:`\theta`
+        wires (Iterable, Wires): the wires the operation acts on
+        control (Wire): The additional ancilla to control the RZ rotation on
+
+    Returns:
+        list[Operator]: decomposition that make up the controlled evolution
+    """
+    ops = [qml.CNOT(wires=(w0, w1)) for w0, w1 in zip(wires[~0:0:-1], wires[~1::-1])]
+    ops.append(qml.CRZ(theta, wires=[control, wires[0]]))
+    ops += [qml.CNOT(wires=(w0, w1)) for w0, w1 in zip(wires[1:], wires[:~0])]
+
+    return ops
+
+
+def evolve_under(ops, coeffs, time):
+    """
+    Evolves under the given Hamiltonian deconstructed into its pauliwords
+
+    Args:
+        ops (List[Observables]): List of Pauliwords that comprise the Hamiltonian
+        coeffs (List[int]): List of the respective coefficients of the Pauliwords of the Hamiltonian
+        time (float): At what time to evaluate these Pauliwords
+    """
+    for op, coeff in zip(ops, coeffs):
+        ControlledPauliEvolution(
+            coeff * time,
+            wires=op.wires,
+            pauli_word=qml.grouping.pauli_word_to_string(op),
+            ancillas=["Hadamard", "Target"],
+        )
+
+
+def calculate_Xi_decomposition(hamiltonian):
+    """
+    Calculates the Xi-decomposition from the given hamiltonian by constructing the sparse matrix representing the hamiltonian, finding its spectrum and then construct projectors and eigenvalue spacings
+
+    Args:
+      hamiltonian (qml.Hamiltonian): The pennylane hamiltonian to be decomposed
+
+    Returns:
+      dEs (List[float]): The step separating the two eigenvalues (E_1-E-2)/2 of the spectrum
+      mus (List[float]): The average between the two eigenvalues (E_1+E-2)/2
+      times (List[float]): The time for this term group to be evaluated/evolved at
+      projs (List[np.array]): The analytical observables associated with these groups, to be measured by qml.Hermitian
+    """
+    mat = qml.utils.sparse_hamiltonian(hamiltonian).toarray()
+    size = len(mat)
+    eigs, eigvecs = np.linalg.eigh(mat)
+    norm = eigs[-1]
+    proj = np.identity(size, dtype="complex64")
+    proj += -2 * np.outer(np.conjugate(eigvecs[:, 0]), eigvecs[:, 0])
+    last_i = 1
+
+    dEs, mus, projs, times = [], [], [], []
+
+    for index in range(len(eigs) - 1):
+        dE = (eigs[index + 1] - eigs[index]) / 2
+        if np.isclose(dE, 0):
+            continue
+        dEs.append(dE)
+        mu = (eigs[index + 1] + eigs[index]) / 2
+        mus.append(mu)
+        time = np.pi / (2 * (norm + abs(mu)))
+        times.append(time)
+
+        for j in range(last_i, index + 1):
+            proj += -2 * np.outer(np.conjugate(eigvecs[:, j]), eigvecs[:, j])
+            last_i = index + 1
```

Comment:
```suggestion
        last_i = index + 1
```
This doesn't depend on `j`, and the line above it doesn't use it.  As far as I can tell, we don't have to update it every time in the inner loop.

Reply:
I agree in principle with what you are guetting at, a few things. 
```
        if np.isclose(dE, 0):
            continue
```

At the beginning of the loop the eigenvalue spacing is checked (to find a degeneracy) and if one is found the outer loop skips a step (and therefore the inner loop) as the contribution for this term would be 0. This construction sums up all the skipped step once a new distinct eigenvalue is found and dE is nonzero. One could move this operation to the top before the check, the advantage of having it here is that if there is a degeneracy at the end of your spectrum you are done and this skips a few iterations of this loop which are not needed, so slightly faster, as things are only summed up when they are really needed